usage_example/Readme.txt

```diff
@@ -0,0 +1,270 @@
+export ARCH=arm64
+export CROSS_COMPILE=aarch64-none-linux-gnu-
+set -e
+----
+
+[Target kernel source:
+Out of tree module compilation requires the Module.symvers file in the 
+target kernel directory. So it is necessary to execute a 'make modules' 
+command as part of Linux kernel compilation/build steps.]
+ 
+(cd to top folder)
+
+
+cd $TOP_FOLDER
+if [ -e linux-socfpga ]
+then
+echo "linux-socfpga exists"
+cd linux-socfpga
+export KERNEL_SRC=`pwd`
+echo "KERNEL_SRC: $KERNEL_SRC"
+cd $TOP_FOLDER
+else
+echo "linux-socfpga does not exist"
+fi
+----
+
+[module compilation, paths need to be changed based on directory structure]
+
+if [ -e contig_mem ]
+then
+echo "contig_mem exists"    
+cd contig_mem/udmabuf
+make clean
+make all
+else
+mkdir contig_mem
+cd contig_mem
+cp -r /proj/work/dmohapatra/contig_mem/udmabuf ./
+cd udmabuf
+make clean
+make all
+fi
+
+==============================================================================
+Kernel configuration and boot-up command:
+
+Depending on the need for size of the contiguous memory buffer(s), 
+appropriate amount of physical memory needs to be reserved at boot up time 
+for later usage. This can be achieved by either reserving memory at boot 
+up time (by using device tree specifications or other mechanisms) or by 
+configuring the Contiguous Memory Allocator 
+mechanism inside the Linux kernel and providing kernel bootup command line 
+parameter to reserve appropriate amount of CMA memory. 
+
+Relevant kernel config options (debug options are not strictly necessary):
+CONFIG_CMA=y
+
+CONFIG_CMA_DEBUG=y
+CONFIG_CMA_DEBUGFS=y
+CONFIG_CMA_SYSFS=y
+CONFIG_CMA_AREAS=19
+CONFIG_DMA_CMA=y
+CONFIG_DMA_PERNUMA_CMA=y
+
+CONFIG_CMA_SIZE_MBYTES=16 (This can be changed appropriately)
+CONFIG_CMA_SIZE_SEL_MBYTES=y
+CONFIG_CMA_ALIGNMENT=8
+CONFIG_DMA_API_DEBUG=y
+CONFIG_DMA_API_DEBUG_SG=y
+
+
+For reserving higher sizes of CMA memory (beyond the 0-4GB physical memory 
+range), following config options need to be turned off.
+
+# CONFIG_ZONE_DMA is not set
+# CONFIG_ZONE_DMA32 is not set
+
+
+
+Settings for other related config options, some of these may be automatically
+switched on based on previous config settings:
+
+CONFIG_HAVE_DMA_CONTIGUOUS=y
+CONFIG_ARCH_HAS_ZONE_DMA_SET=y
+CONFIG_DMA_SHARED_BUFFER=y
+CONFIG_DMA_ENGINE=y
+CONFIG_DMA_VIRTUAL_CHANNELS=y
+CONFIG_DMA_ACPI=y
+CONFIG_DMA_OF=y
+CONFIG_IOMMU_DEFAULT_DMA_LAZY=y
+CONFIG_IOMMU_DMA=y
+CONFIG_HAS_DMA=y
+CONFIG_DMA_OPS=y
+CONFIG_NEED_SG_DMA_FLAGS=y
+CONFIG_NEED_SG_DMA_LENGTH=y
+CONFIG_NEED_DMA_MAP_STATE=y
+CONFIG_ARCH_DMA_ADDR_T_64BIT=y
+CONFIG_DMA_DECLARE_COHERENT=y
+CONFIG_ARCH_HAS_SETUP_DMA_OPS=y
+CONFIG_ARCH_HAS_TEARDOWN_DMA_OPS=y
+CONFIG_ARCH_HAS_SYNC_DMA_FOR_DEVICE=y
+CONFIG_ARCH_HAS_SYNC_DMA_FOR_CPU=y
+CONFIG_ARCH_HAS_DMA_PREP_COHERENT=y
+CONFIG_DMA_BOUNCE_UNALIGNED_KMALLOC=y
+CONFIG_DMA_NONCOHERENT_MMAP=y
+CONFIG_DMA_COHERENT_POOL=y
+CONFIG_DMA_DIRECT_REMAP=y
+
```

Comment:
I am assuming all these are already setup in linux-socfpga and don't need to be turned on.

Reply:
this is all in the linux-socfpga defconfig